Add validator FSM support

flotothemoon opened the pull request:
Adds support for the recently introduced validator registration FSM from Radix engine. Specifically, the API is expanded with two new actions: `RegisterValidatorAction` and `UnregisterValidatorAction` with the corresponding mappers. To reduce logic duplication, validator registration state mapping is implemented in the shared `ValidatorRegistrationState` helper.

Commit:
Add validator particles

flotothemoon committed:
Add validator actions and mappers

Commit:
Add register/unregister action implementation, add example demonstrating their use

Commit:
Add documentation for validator actions and mappers

Commit:
Add test for ValidatorRegistrationState

Commit:
Fix formatting in new validator classes

clstrfsck commented on examples/src/main/java/com/radixdlt/client/examples/ValidatorExample.java:
```diff
@@ -0,0 +1,31 @@
+package com.radixdlt.client.examples;
```
Missing header comment.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java:
```diff
@@ -774,6 +780,38 @@ public Result sendTokens(
+	public Result registerValidator(
+		RadixAddress validator
+	) {
+		Objects.requireNonNull(validator, "validator");
```
Would it be better to do this null check in the public constructor for RegisterValidatorAction?

flotothemoon replied:
That would be nicer, yeah.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java:
```diff
@@ -774,6 +780,38 @@ public Result sendTokens(
+	public Result unregisterValidator(
+		RadixAddress validator
+	) {
+		Objects.requireNonNull(validator, "validator");
```
Similarly here, would it be better to do this null check in the public constructor for UnregisterValidatorAction?

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:
```diff
@@ -0,0 +1,155 @@
+ * The registration state of a certain validator address.
+ */
+final class ValidatorRegistrationState {
+	private final transient Particle particle;
```
Not sure why this is marked "transient" here?

flotothemoon replied:
Brainfart. Removing.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:
```diff
@@ -0,0 +1,155 @@
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
```
Would prefer instanceof check for final class.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	@Override
+	public boolean equals(Object o) {
```
Would like to see an EqualsVerifier unit test for this.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/UnregisteredValidatorParticle.java:
```diff
@@ -0,0 +1,68 @@
+		this.nonce = nonce;
+	}
+
+	public RadixAddress getAddress() {
```
Should have @ Override from Ownable.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java:
```diff
@@ -0,0 +1,68 @@
+		this.nonce = nonce;
+	}
+
+	public RadixAddress getAddress() {
```
Should have @ Override from Ownable.

clstrfsck commented on radixdlt-java/src/test/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationStateTest.java:
```diff
@@ -0,0 +1,156 @@
+package com.radixdlt.client.application.translate.validators;
```
Missing header comments.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java:
```diff
@@ -0,0 +1,39 @@
+package com.radixdlt.client.application.translate.validators;
```
Missing header comment.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorAction.java:
```diff
@@ -0,0 +1,25 @@
+package com.radixdlt.client.application.translate.validators;
```
Missing header comment.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorAction.java:
```diff
@@ -0,0 +1,25 @@
+package com.radixdlt.client.application.translate.validators;
```
Missing header comment.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java:
```diff
@@ -0,0 +1,39 @@
+package com.radixdlt.client.application.translate.validators;
```
Missing header comment.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:
```diff
@@ -0,0 +1,155 @@
+package com.radixdlt.client.application.translate.validators;
```
Missing header comment.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/RegisteredValidatorParticle.java:
```diff
@@ -0,0 +1,68 @@
+/*
```
I think this is the wrong license header for the library.  Should be the MIT header.

flotothemoon replied:
Indeed, will replace.

clstrfsck commented on radixdlt-java/src/main/java/com/radixdlt/client/atommodel/validators/UnregisteredValidatorParticle.java:
```diff
@@ -0,0 +1,68 @@
+/*
```
This is the wrong header comment for the library.  Should be the MIT header.

clstrfsck reviewed (changes requested):
A couple of missing header comments and some other items that could be considered.

flotothemoon committed:
Add missing/correct file headers

Commit:
Move validator particle static parameter check to constructors

Commit:
Add equalsverifier test for ValidatorRegistrationState

talekhinezh commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/RegisterValidatorActionMapper.java:
```diff
@@ -0,0 +1,61 @@
+	public Set<ShardedParticleStateId> requiredState(RegisterValidatorAction action) {
+		return ImmutableSet.of(
+			ShardedParticleStateId.of(RegisteredValidatorParticle.class, action.getValidator()),
+			ShardedParticleStateId.of(UnregisteredValidatorParticle.class, action.getValidator())
```
I would think you only need UnregisteredValidatorParticle here no?

flotothemoon replied:
Leaving for now as discussed

talekhinezh commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/UnregisterValidatorActionMapper.java:
```diff
@@ -0,0 +1,61 @@
+	public Set<ShardedParticleStateId> requiredState(UnregisterValidatorAction action) {
+		return ImmutableSet.of(
+			ShardedParticleStateId.of(RegisteredValidatorParticle.class, action.getValidator()),
+			ShardedParticleStateId.of(UnregisteredValidatorParticle.class, action.getValidator())
```
I would think you only need RegisteredValidatorParticle here no?

flotothemoon replied:
Leaving for now as discussed

talekhinezh commented on radixdlt-java/src/main/java/com/radixdlt/client/application/translate/validators/ValidatorRegistrationState.java:
```diff
@@ -0,0 +1,177 @@
+/**
+ * The registration state of a certain validator address.
+ */
+final class ValidatorRegistrationState {
```
Can we rename this as its easy to confuse with it being an ApplicationState

flotothemoon replied:
True, though it being local to the validators package makes that somewhat less worrisome. Happy to rename anyway, do you have suggestions?

Commit:
Fix equalsverifier test for ValidatorRegistrationState

Prefab instances of abstract class "Particle" all serialize to the same empty data, and therefore are always equals(...) and have the same hashCode().
